src/schema/feeds.ts

```diff
@@ -686,6 +688,66 @@ const applyFeedPaging = (
   return newBuilder;
 };
 
+const feedPageGeneratorWithPin: PageGenerator<
+  GQLPost,
+  FeedArgs,
+  FeedPage,
+  unknown
+> = {
+  connArgsToPage: ({ first, after }: FeedArgs) => {
+    const cursor = getCursorFromAfter(after);
+    const limit = Math.min(first || 30, 50) + 1;
+
+    return cursor
+      ? { limit, timestamp: new Date(parseInt(cursor)) }
+      : { limit };
+  },
+  nodeToCursor: (_, __, node) => {
+    const createdAt = new Date(node.createdAt);
+    const timestamp =
+      node.type === PostType.Welcome
+        ? changeYearToNextYear(createdAt)
+        : createdAt;
+
+    return base64(`time:${new Date(timestamp).getTime()}`);
+  },
+  hasNextPage: (page, nodesSize) => page.limit === nodesSize,
+  hasPreviousPage: (page) => !!page.timestamp,
+  transformNodes: (page, nodes) => nodes.slice(0, page.limit - 1),
+};
+
+const applyFeedPagingWithPin = (
+  ctx: Context,
+  page: FeedPage,
+  builder: SelectQueryBuilder<Post>,
+  alias,
+): SelectQueryBuilder<Post> => {
+  const priority = `CASE WHEN ${alias}.type = '${PostType.Welcome}' THEN 1 ELSE 0 END`;
+  let newBuilder = builder
+    .addSelect(`${alias}."createdAt" as "createdAt"`)
+    .addSelect(priority, 'priority')
+    .limit(page.limit)
+    .orderBy(`priority`, 'DESC')
+    .addOrderBy(`${alias}."createdAt"`, 'DESC');
+
+  if (page.timestamp) {
+    // to ensure pinned post (or welcome post) be pinned on top of the feed, we need to ensure they have greater date than the rest of the result
+    // that is solved by making the year of pinned posts to the next year and since we retained the other parts of the date, sorting should be accurate.
+    // the current condition is about being a welcome post, but on v6 this will be based on the `pinned` property.
+    const nextYearDifference = `(DATE_PART('year', CURRENT_DATE) - DATE_PART('year', ${alias}."createdAt") + 1)`;
+    newBuilder = newBuilder.andWhere(
```

Comment:
My only concern here is whether this query still uses the `createdAt` index we have. But I guess we can check it after we deploy. But we need to be mindful to the fact it may ignore indexes

Reply:
Good point. Let's closely monitor the performance difference vs the previous one once merged.

Reply:
Also, the current implementation forces ranking by `time`, which is basically our FE implementation. Would you want to open the opportunity to request for `popularity` based for source feed query? That will be another formula to make for adjusting the scoring system. Or we can tackle it when the time comes?

Reply:
We can tackle when the time comes. Hopefully we will migrate this responsibility to the feed service anyway